portico: Add script to auto-generate message screenshots /for/research.

roanster007 opened the pull request:
Previously, we manually generated screenshots on /for/research and other landing pages, which makes them difficult to maintain.

This PR adds a script - generate-user-messages-screenshot, which expects the path of the filename containing the thread conversation and path where screenshot needs to be generated, and generates screenshot at the mentioned location.

Fixes #30016

<!-- Describe your pull request here.-->


<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

Testing: `./tools/screenshots/generate-user-messages-screenshot --thread tools/screenshots/interactive_messaging.json`

**Screenshots and screen captures:**

![image](https://github.com/zulip/zulip/assets/97049524/16209c66-d4f7-42a2-ace4-be091c2fc1ec)
<br>
![image](https://github.com/zulip/zulip/assets/97049524/79739662-78e9-4889-a853-51fe407ef6b7)
<br>
![image](https://github.com/zulip/zulip/assets/97049524/c2fb9c9c-4b36-489f-ba05-a17619131246)

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

roanster007 commented on tools/generate-user-messages-screenshot:
```diff
@@ -0,0 +1,197 @@
+    access_message(user_profile=DEFAULT_USER, message_id=message_id)
+    for message_id in message_thread_ids
+]
+do_delete_messages(realm, thread_messages)
```
I am deleting all the messages added after the process ends because next time when you re run the script with the same data, then the previous messages would still be visible.

timabbott replied:
Yeah, that's pretty reasonable.

roanster007 replied:
done!

timabbott commented on tools/generate-user-messages-screenshot:
```diff
@@ -0,0 +1,197 @@
+
+realm = get_realm("zulip")
+with open(options.thread[0]) as f:
+    threads = json.load(f)
```
What does the configuration file look like? I'm not sure JSON is likely to be the most convenient format for drafting these; could imagine just using a set of Python dataclasses instead, each one declared in an individual file in a `tools/screenshots/` directory ... which probably we should create and move the existing tool into as a prep step for this PR...

timabbott replied:
Oh, I see, it's the next commit; yeah let's at least do the directory structure change proposed in my comment to give this cluster of tools its own directory to avoid cluttering the `tools/` root with at least one file per screenshot, and then perhaps @alya can review and consider whether she's prefer a different format for composing these.

roanster007 replied:
updated!

timabbott commented:
Overall, this looks great; posted a few comments but I'm generally in favor of merging this sort of thing and then doing iterative refinements on top of it.

roanster007 commented:
>Better would be to use try/finally to ensure they get cleaned up even if the script crashes.

I updated to clean up the messages added in the `finally` block.

>Oh, I see, it's the next commit; yeah let's at least do the directory structure change proposed in my comment to give this cluster of tools its own directory to avoid cluttering the tools/ root with at least one file per screenshot

- I moved `generate-integration-docs-screenshots` and `message-screenshot` to `tools/screenshots` directory as prep commit.
- Added the new script files there as well.

Should the data for the message conversation (`interactive_messaging.json`) also be moved to this dir? Currently I added it in `static/images/landing-page/research/interactive_messaging.json` since that is where the current images for `for/research` page reside.

alya commented on static/images/landing-page/research/interactive_messaging.json:
```diff
@@ -0,0 +1,90 @@
+            },
+            {
+                "sender": "Zoe Davis",
+                "content": "@**everyone** Video call is happening now! [Click to join video call](#)"
```
Let's change [Click to join video call] -> [Join video call.], since we changed the default text the button inserts.

roanster007 replied:
updated!

alya commented:
Let's make the times more realistic, so that it doesn't look like all the messages were sent at the same time. I think the ones in the current screenshots are basically fine.

alya commented:
"Today" is generally good for the date, and thanks for updating the global time. But let's make the storyline be that the call gets scheduled the day before, and then the "happening now" message is sent at the time that the call is starting.

alya commented:
@terpimost Could you please share images we should use for avatars for these conversations?

alya commented on static/images/landing-page/research/interactive_messaging.json:
```diff
@@ -0,0 +1,90 @@
+                "content": "Hey! How does <time:2024-05-10T17:00:00+05:30> sound for doing a video call today? My paper just got rejected.",
+                "reactions": {
+                    "+1": ["James Williams", "Iago", "Zoe Davis", "John Doe", "Abraham Lincoln"],
+                    "broken_heart": ["James Williams", "Iago", "Zoe Davis", "Abraham Lincoln"]
```
It doesn't matter (since it's not visible in the screenshot), but it might be a bit nicer to use character names from https://zulip.com/ than Abraham Lincoln, etc.?

roanster007 replied:
😅it was the most random thing I came up with! I'll update it.

timabbott commented:
> Should the data for the message conversation (interactive_messaging.json) also be moved to this dir? Currently I added it in static/images/landing-page/research/interactive_messaging.json since that is where the current images for for/research page reside.

Yeah, let's have it live in this directory; it's good to have all the source materials in one place.

timabbott commented on tools/screenshots/generate-integration-docs-screenshot:
```diff
@@ -225,7 +226,7 @@ def capture_last_message_screenshot(bot: UserProfile, image_path: str) -> None:
         return
     message_id = str(message.id)
-    screenshot_script = os.path.join(TOOLS_DIR, "message-screenshot.js")
+    screenshot_script = os.path.join(SCREENSHOTS_DIR, "message-screenshot.js")
```
Never rename something without doing a `git grep` to see if it's referenced elsewhere:

```
$ git grep generate-integration-docs-scree
api_docs/incoming-webhooks-overview.md:  generated by running `tools/generate-integration-docs-screenshot --integration mywebhook`.
api_docs/incoming-webhooks-overview.md:  `tools/generate-integration-docs-screenshot --integration mywebhook`.
docs/documentation/integrations.md:  the screenshot using `tools/generate-integration-docs-screenshot`:
docs/documentation/integrations.md:  ./tools/generate-integration-docs-screenshot --integration integrationname
zerver/tests/test_integrations.py:            'and running "tools/generate-integration-docs-screenshot" to keep the screenshots up-to-date.'
```

You need to update those callers.

roanster007 replied:
updated

timabbott commented:
We can probably not block on finalizing the content of the `json` file and what avatars to use for merging the tool, so I'm good with merging this PR as soon as my comments above are resolved, and the rest will probably feel nice to do as follow-ups, but of course it'd be great to check off as many things as we can from Alya's feedback.

roanster007 commented on tools/screenshots/generate-user-messages-screenshot:
```diff
@@ -0,0 +1,200 @@
+
+finally:
+    thread_messages = [
+        access_message(user_profile=DEFAULT_USER, message_id=message_id)
```
`DEFAULT_USER` (iago) is declared at global scope, so I guess we should be able to access it here.

timabbott replied:
No, I mean if `access_message` fails. I guess it shouldn't as long as the current user isn't a guest; it just seems weird to query this way.

roanster007 replied:
ya makes sense! updated!

roanster007 commented:
@alya addressed reviews:
>It doesn't matter (since it's not visible in the screenshot), but it might be a bit nicer to use character names from https://zulip.com/ than Abraham Lincoln, etc.?

>Let's change [Click to join video call] -> [Join video call.], since we changed the default text the button inserts.

> Let's make the times more realistic, so that it doesn't look like all the messages were sent at the same time. I think the times in the screenshots that are currently on the website are fine, but I'm not super picky.

>  "Today" is generally good for the date, and thanks for updating the global time. But let's make the storyline be that the call gets scheduled the day before, and then the "happening now" message is sent at the time that the call was scheduled for.

And updated screenshots accordingly.

roanster007 commented:
@timabbott 
> Yeah, let's have it live in this directory; it's good to have all the source materials in one place.

changed the location of the `json` file to `tools/screenshots.

> No, I mean if access_message fails. I guess it shouldn't as long as the current user isn't a guest; it just seems weird to query this way instead of Message.objects.filter(id__in=thread_message_ids)

Removed `access_message` and updated with this.

> Never rename something without doing a git grep to see if it's referenced elsewhere:

Updated `tools/generate-integration-docs-screenshots` to `tools/screenshots/generate-integration-docs-screenshots`.

- Added a preparatory commit so as to be able to send messages with `forged_timestamp` using `internal_prep_stream_message`

alya commented:
Works for me!

timabbott commented:
Merged, thanks @roanster007!

timabbott commented:
A few follow-up requests:
- Can we disable the log output for sending outgoing emails in the tool, and give it proper usage output, to make it a bit more pleasant to use?
- Can you document this on localhost:9991/devtools?
- Probably you can start working through writing the JSON files for other screenshots on /for/* landing pages.

alya commented:
> Probably you can start working through writing the JSON files for other screenshots on /for/* landing pages.

Please skip /for/business for now; I need to do some content work there.

roanster007 commented:
I'll update it
